rali/rali/source/tf_record_reader.cpp

```diff
@@ -174,106 +187,113 @@ void TFRecordReader::replicate_last_image_to_fill_last_shard()
 
+Reader::Status TFRecordReader::tf_record_reader()
+{
+    std::string fname = _folder_path;
+    // if _record_name_prefix is specified, read only the records with prefix
+    if  (_record_name_prefix.empty() || fname.find(_record_name_prefix) != std::string::npos) {
+        uint file_size;
+        std::ifstream file_contents(fname.c_str(), std::ios::binary);
+        if (!file_contents)
+            THROW("TFRecordReader: Failed to open file " + fname);
+        file_contents.seekg(0, std::ifstream::end);
+        file_size = file_contents.tellg();
+        // std::cerr<<"\n length of the file:: "<<length<<std::endl;
+        file_contents.seekg(0, std::ifstream::beg);
+        auto ret = read_image_names(file_contents, file_size);
+        if (ret != Reader::Status::OK)
+            THROW("TFRecordReader: Error in reading TF records");
+        _last_rec = false;
+        if (_file_names.size() != _file_size.size())
+            std::cerr << "\n Size of vectors are not same";
+        file_contents.close();
+    }
     return Reader::Status::OK;
 }
 
 size_t TFRecordReader::get_file_shard_id()
 {
-    if(_batch_count == 0 || _shard_count == 0)
+    if (_batch_count == 0 || _shard_count == 0)
         THROW("Shard (Batch) size cannot be set to 0")
-    return (_file_id / (_batch_count)) % _shard_count;
+    return _file_id  % _shard_count;
 }
 
-
 Reader::Status TFRecordReader::read_image_names(std::ifstream &file_contents, uint file_size)
 {
     auto ret = Reader::Status::OK;
-    uint length;
-    size_t uint64_size, uint32_size;
-    uint64_t data_length;
-    uint32_t length_crc, data_crc;
-    uint64_size = sizeof(uint64_t); 
-    uint32_size = sizeof(uint32_t);
-    while(!_last_rec)
+    while (!_last_rec)
     {
+        uint length;
+        uint64_t data_length;
+        uint32_t length_crc, data_crc;
+
         length = file_contents.tellg();
-        char *header_length = new char[uint64_size];
-        char *header_crc = new char[uint32_size];
-        char *footer_crc = new char[uint32_size];
-        file_contents.read(header_length, uint64_size);
+        file_contents.read((char *)&data_length, sizeof(data_length));
         if(!file_contents)
             THROW("TFRecordReader: Error in reading TF records")
-        file_contents.read(header_crc, uint32_size);
+        file_contents.read((char *)&length_crc, sizeof(length_crc));
         if(!file_contents)
             THROW("TFRecordReader: Error in reading TF records")
-        memcpy(&data_length, header_length, sizeof(data_length));
-        memcpy(&length_crc, header_crc, sizeof(length_crc));
-        if(uint(length + data_length + 16) == file_size)
+        if (uint(length + data_length + 16) == file_size)
         {
             _last_rec = true;
-
         }
-        char *data = new char[data_length];
-        file_contents.read(data,data_length);
+        std::unique_ptr<char[]> data(new char[data_length]);
+        file_contents.read(data.get(), data_length);
```

Comment:
Codacy found an issue: [Check buffer boundaries if used in a loop including recursive loops (CWE-120, CWE-20).  ](https://app.codacy.com/gh/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/pullRequest?prid=6366586)